### PR TITLE
Fix the problem of output outliving the pipeline in python

### PR DESCRIPTION
**Pull request by JanuszL**
- fixes the problem when the Python gcs pipeline object but there are still outputs that reference to the underlying memory

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It fixes the problem when the Python gcs pipeline object but there are still outputs that reference to the underlying memory

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     bound lifetime of the pipeline to the returned outputs
 - Affected modules and functionalities:
     backend_impl
 - Key points relevant for the review:
     NA
 - Validation and testing:
     test case has been added
 - Documentation (including examples):
     NA

**JIRA TASK**: *[DALI-1657]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/python/backend_impl.cc [hunk 1]
```diff
@@ -1248,34 +1248,56 @@ PYBIND11_MODULE(backend_impl, m) {
...
     .def("RunGPU", &Pipeline::RunGPU)
     .def("Outputs",
-        [](Pipeline *p) {
+        [](py::object &p_o) {
+          Pipeline *p = py::cast<Pipeline *>(p_o);
           DeviceWorkspace ws;
           p->Outputs(&ws);
 
-          py::list list;
+          py::tuple tup(ws.NumOutput());
```
Comment (jantonguirao) [inline dali/python/backend_impl.cc:1256]:
```suggestion
          py::tuple outs(ws.NumOutput());
```

**Review by jantonguirao — APPROVED** [review]

File: dali/python/backend_impl.cc [hunk 2]
```diff
@@ -1248,34 +1248,56 @@ PYBIND11_MODULE(backend_impl, m) {
...
-              list.append(&ws.Output<CPUBackend>(i));
+              tup[i] = &ws.Output<CPUBackend>(i);
+              // it is not possible to obtain weekref to the tuple and bound it to the pipeline
+              // object to keep the pipeline object alive as long as a returned tuple is valid,
+              // so instead keep pipeline alive as long as any element of the returned
+              // tuple is alive
+              py::detail::keep_alive_impl(tup[i], p_o);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              tup[i] = &ws.Output<GPUBackend>(i);
```
Comment (jantonguirao) [inline dali/python/backend_impl.cc:1292]: you can move this outside of the if and have it only once

File: dali/python/backend_impl.cc [hunk 3]
```diff
@@ -1248,34 +1248,56 @@ PYBIND11_MODULE(backend_impl, m) {
...
           DeviceWorkspace ws;
           p->Outputs(&ws);
 
-          py::list list;
+          py::tuple tup(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              tup[i] = &ws.Output<CPUBackend>(i);
+              // it is not possible to obtain weekref to the tuple and bound it to the pipeline
```
Comment (jantonguirao) [inline dali/python/backend_impl.cc:1260]:
```suggestion
              // it is not possible to obtain weakref to the tuple and bound it to the pipeline
```

Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

File: dali/python/backend_impl.cc (hunk 2 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

File: dali/python/backend_impl.cc (hunk 1 above)
Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

**Comment by JanuszL** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Review by mzient — APPROVED** [review]

File: dali/python/backend_impl.cc [hunk 4]
```diff
@@ -1248,34 +1248,46 @@ PYBIND11_MODULE(backend_impl, m) {
...
+          py::tuple outs(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              outs[i] = &ws.Output<CPUBackend>(i);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              outs[i] = &ws.Output<GPUBackend>(i);
             }
+            // it is not possible to obtain weakref  to the tuple and bound it to the pipeline
```
Comment (klecki) [inline dali/python/backend_impl.cc:1284]:
```suggestion
            // it is not possible to obtain weakref to the tuple and bind it to the pipeline
```
Same above.

Btw, why did you change it to tuple?

Comment (JanuszL) [reply dali/python/backend_impl.cc:1284]: We have returned it as tuple anyway `return  py::cast<py::tuple>(list);`.

Comment (JanuszL) [reply dali/python/backend_impl.cc]: Done

**Comment by mzient** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

File: dali/test/python/test_pipeline.py [hunk 5]
```diff
@@ -1696,3 +1696,14 @@ def define_graph(self):
...
+def test_pipeline_out_of_scope():
+    def get_output():
+        pipe = dali.pipeline.Pipeline(1, 1, 0)
+        with pipe:
+            pipe.set_outputs(dali.fn.uniform())
+        pipe.build()
+        return pipe.run()
+    data = get_output()
+    out = data[0].as_array()[0]
+    print(out)
```
Comment (mzient) [inline dali/test/python/test_pipeline.py:1709]:
```suggestion
            pipe.set_outputs(dali.fn.external_source(source=[[np.array([-0.5, 1.25])]]))
        pipe.build()
        return pipe.run()[0]
    out = get_output().at(0)
    assert out[0] == -0.5 and out[1] == 1.25
```

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!builld

**Review by jantonguirao — APPROVED** [review]
